### ISSUE TITLE
Tags, routes, and other configuration resets on reboot (sometimes)

### ISSUE THREAD

**Issue by gregg098** [tailscale/tailscale#2258]
My Tailscale network consists of about 12 devices. About half of them have some kind of custom configuration including tags or other parameters. For example:

My DNS machine is configured like this:
`sudo tailscale up --accept-routes=false --accept-dns=false --advertise-tags=tag:dns`

Some of my VPS machines are configured like this:
`sudo tailscale up --advertise-exit-node --accept-routes=false --accept-dns=false --advertise-tags=tag:vps`

Thave I have one subnet node, something like this (I changed the route a little to simplify the example):
`sudo tailscale up --advertise-routes=192.168.10.0/24,192.168.20.0/24 --advertise-exit-node --accept-routes=false --accept-dns=false`

Outside of clients with no custom configuration, most of the machines are Debian/Ubuntu-based, but there is one Windows Server in the mix. I have not tested the Windows machine too much yet, but I think its pretty stable.

The problem is I have is that I randomly lose tags, routes, etc. on a reboot of the Linux machines. This can be just a normal reboot and not associated with any updates.  I can't isolate the exact cause, but various steps seem to help (most of the time).

When this happens and I notice it (usually because something isn't routing properly), I have to do one of two things to fix it. 
1) `sudo tailscale up --flags --reset`
or
2) `sudo tailscale down` and then `sudo tailscale up --flags` In this case, I have run into instances where I had to pass the --reset flag to see the tags and routes back in the dashboard.

Simply giving it the `sudo tailscale up --flags` or even `sudo tailscale up` doesn't always work. It will start the service, but no tags/routes are passed.

From testing, it almost seems like it "forgets" that it has custom flags, but again, its not consistent so I'm not sure.

Yesterday, it seemed like if I did the following steps on one of the problem machines, the problem went away, but then I had issues on another machine that didn't reboot or anything.
1) `sudo tailscale down`
2) `sudo tailscale up --flags --reset`
3) `sudo tailscale up --flags`

Again, maybe way off base, but it seems like if I have to use `--reset` that it works temporarily, but then resets to almost a default state if the machine restarts. So passing the same command line a second time without `--reset` lets a reboot succeed.

Let me know if there is a way to supply any logs when I see this happen so we can track it down. I may attempt to fully remove Tailscale (including `/var/lib/tailscale/tailscaled.state`) and re-setup some of the machines in the next day or two if nothing else works.

Additional information:
- Over the last week or so I have modified my configuration quite a bit through experimenting. So flags have changed a few times and the command line suggests that I use `--reset` when I do this.
- I do have ACLs setup to somewhat isolate the `tag:vps` machines so they can only talk to themselves. This shouldn't impact tagging and routes from what I know of.
- The Linux machines are a mix of KVM, LXC, and Proxmox host nodes.

<img src="https://frontapp.com/assets/img/favicons/favicon-32x32.png" height="16" width="16" alt="Front logo" /> [Front conversations](https://app.frontapp.com/open/top_3b2oh)

**Comment by bradfitz** [comment]
Which version of Tailscale?

Which Linux distro & version?

If this happens again (or you can force it to happen), run `tailscale bugreport` and share its output so we can then go look at logs around that time.

**Comment by gregg098** [comment]
Its a mix of machines, all Debian based.
A few Debain 9 LXC's
One might be Debian 10
Two Ubuntu 20.04 KVMs

I just rebooted a few of them and none of them lost their tags or routes. I think that the steps I called out above seem to work (for now).

1) `sudo tailscale down`
2) `sudo tailscale --flags --reset`
3) `sudo tailscale --flags` (same as above but without --reset - No idea if this is necessary, but everything works right now)

Will use the bugreport command if it happens again and post a reply here.

Thanks.

**Comment by gregg098** [comment]
I take that back. Two of the machines were actually showing up as offline (in the Tailscale dashboard). They are fully accessible locally. Both are PiHole servers and continue to serve DNS records to the rest of my home network apprpriately.

Bug report from each:
BUG-1e926521a935546116cb72e1c1bf9dcf91b272d7041ccc54288f08a039b1e27d-20210627190923Z-e2c908b0761704d6

and

BUG-76285545edf6ae463d6db17805f7a033ed84e3ba132cbe4234b1770cb24ef8e5-20210627191035Z-a3630d6df5dbf40d

Both bug reports are after the reboot, but before re-issueing sudo tailscale up --flags

**Comment by gregg098** [comment]
More trivia. For one of the PiHole nodes that I just issued a bug report on. I was able to get it to reboot without having to mess with re-issuing tailscale up commands.
This is the normal command for that machine:
`sudo tailscale up --accept-routes=false --accept-dns=false --advertise-tags=tag:dns`

This time, I added `--force-reauth` and went through re-authorization. After a reboot, it kept its tag.

Another note, most of the machines with issues have no expiration on keys set. I dont know if thats related or not.

**Comment by bradfitz** [comment]
Do you have shutdown scripts trying to call `tailscale down` prior to reboot?

Because `tailscale down` records your intention to not be running which persists after reboots, which looks like what I'm seeing in logs at least in some of your cases I've looked at so far.

**Comment by gregg098** [comment]
> Do you have shutdown scripts trying to call `tailscale down` prior to reboot?
> 
> Because `tailscale down` records your intention to not be running which persists after reboots, which looks like what I'm seeing in logs at least in some of your cases I've looked at so far.

No. The LXC only runs PiHole and was rebooted with sudo reboot. As far as I know tailscale is running just fine before the reboot.

**Comment by bradfitz** [comment]
@gregg098, when it starts back up, I see in your logs it starting in stopped state, because of a previous "tailscale down". I can't reproduce it improperly getting stuck in that state on its own.

Can you try this:

* verify it's running, with `tailscale status`. You can also run `tailscale debug -prefs` and observe `"WantRunning": true,`.
* take a bug report to get a bug report log marker (`tailscale bugreport` and share the BUG-... line)
* reboot the machine
* do another `tailscale bugreport` and check the status.

Basically, I'd like to see a pair of bug report markers before & after the reboot, with no `tailscale down` involved.

**Comment by gregg098** [comment]
I tried three machines this morning and none of them had any issues recovering from a reboot. 
Here are the pre-reboot (w/ Tailscale verified running) bugreports if you want them, but they I'm guessing they wont show anything this time.
BUG-1e926521a935546116cb72e1c1bf9dcf91b272d7041ccc54288f08a039b1e27d-20210628153708Z-045a0dbd23107037
BUG-76285545edf6ae463d6db17805f7a033ed84e3ba132cbe4234b1770cb24ef8e5-20210628153902Z-cbfc86d3fba2a12e
BUG-ea289f1898846353be6a328339c7c9cc67d3a059a429b00ce8a8d3fa1181db27-20210628154001Z-7ff65c15d8c51e2d

I haven't had a single issue since I re-issued the tailscale up command with --force-reauth and re-signing in yesterday. I wonder if it was related to a key set to not expire and a config that changed quite a bit over the last week while testing things out? Ill keep an eye on things the next few days and reply if I have any more issues. Maybe Ill do some random reboots here and there just to test.

**Comment by sumocomputers** [comment]
Just wanted to piggyback on @gregg098 since his use of `--force-reauth` has solved an issue for me that nobody could resolve, including Tailscale support.  My typical command would be:


My issue was that when rebooting my Raspberry Pis, Tailscale would never autostart. `tailscale status` would always return stopped. My Pis also run Pi-Hole.

My typical command would be

**Comment by gregg098** [comment]
Checking back in. No issues in awhile now. Ever since I used --force-reauth
Today I updated my servers and rebooted every single one of them and everything came up perfectly.

**Comment by gregg098** [comment]
Adding a new bugreport to the mix. Added a new test machine and had some very similar issues as listed above. Its an Ubuntu 20.04 KVM instance on a VPS. Got tailscale setup and running with this command:
`sudo tailscale up --advertise-exit-node --accept-routes=false --accept-dns=false --advertise-tags=tag:vps`

Accepted the exit node and set the key to never expire. In the last day or so I have been unable to reach that machine from my Windows laptop through Tailscale. A `tailscale ping node` just times out. If I SSH to the new machine (directly), then `tailscale ping laptop` it gets a pong instantly. If I then try from my laptop again, it works just fine. Wait and hour or two and then same thing happens again.

I went ahead and issued the `tailscale up` command above but with `--reset` at the end. Then again with `--force-reauth` but not `--reset`
This has fixed issues every other time that were similar.

I have nothing to back this up, but it almost seems like setting the key to not expiring introduces the issue. Then forcing that machine to reset and reauth updates everything to normal.

Here is a bug report from right after the reset and reauth commands. I forgot to do it right before. If I have more issues Ill generate a new one.

BUG-8c58770b036650965694d2dc65950a44914c063d11cee438a1384b26b4584fc1-20210710030314Z-5ea2dd78d6663185

**Comment by bradfitz** [comment]
This was fixed in the just-released 1.10.2. There many dups of this bug but the one where the clues came together for us was in #2321.

Before upgrading to 1.10.2, if you run `tailscale down; tailscale up` it'll force saving the prefs to disk so they won't be lost on upgrade.
